### DISCUSSION TITLE
How to get S3 bucket region.

### DISCUSSION THREAD

**Discussion by ShimYama** [aws/aws-sdk-cpp#2623] [Q&A] [answered]
If I want to get region of a bucket, I can use S3::S3Client::GetBucketLocation method.
But [aws api reference](https://docs.aws.amazon.com/AmazonS3/latest/API/API_GetBucketLocation.html) says we should use HeadBucket method.

> We recommend that you use [HeadBucket](https://docs.aws.amazon.com/AmazonS3/latest/API/API_HeadBucket.html) to return the Region that a bucket resides in. For backward compatibility, Amazon S3 continues to support GetBucketLocation.

But S3::S3Client::HeadBucket method doesn't return any information about bucket region...

GetBucketLocation method has another problem. If the bucket region is 'us-east-1', the result of S3::S3Client::GetBucketLocation is NOT_SET.  It sounds very strange...

Is there any smarter ways to get bucket region?

**Comment by jmklix** [comment]
It looks like this sdk isn't returning the headers when there is `NoResult`. The headers contain the region info that you are looking for. This is a bug, so we're looking into fixing this.

`us-east-1` is [special](https://docs.aws.amazon.com/AmazonS3/latest/API/API_GetBucketLocation.html):
> Buckets in Region us-east-1 have a LocationConstraint of null

This is expected behavior. I believe it is this way because all buckets used to not be associated to any region

**Reply by ShimYama** [reply]
Thank you for your answer.

> It looks like this sdk isn't returning the headers when there is NoResult. The headers contain the region info that you are looking for. This is a bug, so we're looking into fixing

When the bug is fixed, should I use HeadBucket API rather than GetBucketLocation API that aws api reference reccomends?

> us-east-1 is [special](https://docs.aws.amazon.com/AmazonS3/latest/API/API_GetBucketLocation.html):
> Buckets in Region us-east-1 have a LocationConstraint of null
> This is expected behavior. I believe it is this way because all buckets used to not be associated to any region

I understood.
However, using this method with `Aws::S3::Model::BucketLocationConstraintMapper::GetNameForBucketLocationConstraint` is a bit incovinient to me.

I wrote these source to find bucket location before upload and I have to write ridiculously long ternary operator line for us_east_1 exception...

```c++
Aws::S3::S3Client s3client;
Aws::S3::Model::GetBucketLocationRequest request;
request.SetBucket("<Some Bucket Name>");
Aws::String bucketLocationStr;

const auto outcome = s3client.GetBucketLocation(request);
if (outcome.IsSuccess()) {
   const auto location = outcome.GetResult().GetLocationConstraint();
   bucketLocationStr = Aws::S3::Model::BucketLocationConstraintMapper::GetNameForBucketLocationConstraint(location != Aws::S3::Model::BucketLocationConstraint::NOT_SET ? location : Aws::S3::Model::BucketLocationConstraint::us_east_1);
}
```

What do you think?

**Reply by fyse-nassar** [reply]
> It looks like this sdk isn't returning the headers when there is NoResult. The headers contain the region info that you are looking for. This is a bug, so we're looking into fixing this.

@jmklix, is it fixed?

**Reply by sbiscigl** [reply, marked as answer]
this was fixed in [1.11.212](https://github.com/aws/aws-sdk-cpp/releases/tag/1.11.212) when we merged in customizations for express zonal one alongside other changes for re-invent last year. there are several important bug fixes related to it though so i would suggest picking up from HEAD. example code:

```cpp
#include <aws/core/Aws.h>
#include <aws/core/utils/logging/LogLevel.h>
#include <aws/s3/S3Client.h>
#include <aws/s3/model/HeadBucketRequest.h>

using namespace Aws;
using namespace Aws::S3;
using namespace Aws::S3::Model;
using namespace Aws::Utils::Logging;

int main() {
    SDKOptions options;
    options.loggingOptions.logLevel = LogLevel::Debug;
    InitAPI(options);
    {
        S3Client client;
        const auto headBucket = client.HeadBucket(HeadBucketRequest().WithBucket("BucketName"));
        assert(headBucket.IsSuccess());
        std::cout << headBucket.GetResult().GetBucketRegion() << "\n";
    }
    ShutdownAPI(options);
    return 0;
}
```
